From a64730842fa2eed6aff6b3d39682a3f98ea7cf02 Mon Sep 17 00:00:00 2001
From: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Date: Tue, 23 Apr 2024 09:35:49 +0200
Subject: [PATCH] rpmsg: irq: probe driver earlier to fix device init issue

If irq rpmsg driver is probed after i2c rpmsg driver then the
following error message occurs:
rpmsg_i2c 2.i2c: Failed to create device link (0x180)
with 1.interrupt-controller
Indeed i2c rpmsg needs probe of irq rpmsg for interrupt
controller initialization
So probe irq rpmsg driver at the earliest.

Change-Id: I36dde83d89cd03f2c36bfb6bd409ea817fc81a84
Signed-off-by: Gwenael Treuveur <gwenael.treuveur@foss.st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/st/linux-stm32/+/375874
Reviewed-by: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Reviewed-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
Domain-Review: Arnaud POULIQUEN <arnaud.pouliquen@st.com>
Tested-by: Gwenael TREUVEUR <gwenael.treuveur@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
---
 drivers/irqchip/irq-rpmsg.c | 14 +++++++++++++-
 1 file changed, 13 insertions(+), 1 deletion(-)

diff --git a/drivers/irqchip/irq-rpmsg.c b/drivers/irqchip/irq-rpmsg.c
index a7b2bbf3e75c..37db2ddfca50 100644
--- a/drivers/irqchip/irq-rpmsg.c
+++ b/drivers/irqchip/irq-rpmsg.c
@@ -302,7 +302,19 @@ static struct platform_driver irq_rpmsg_plat_driver = {
 		.of_match_table	= of_match_ptr(irq_rpmsg_plat_ids),
 	}
 };
-module_platform_driver(irq_rpmsg_plat_driver);
+
+static int __init irq_rpmsg_init(void)
+{
+	return platform_driver_register(&irq_rpmsg_plat_driver);
+}
+
+static void __exit irq_rpmsg_exit(void)
+{
+	platform_driver_unregister(&irq_rpmsg_plat_driver);
+}
+
+subsys_initcall(irq_rpmsg_init);
+module_exit(irq_rpmsg_exit);
 
 MODULE_DESCRIPTION("RPMsg IRQC Driver");
 MODULE_LICENSE("GPL");
-- 
2.39.5

